go/vt/vtgate/planbuilder/operators/horizon_expanding.go

```diff
@@ -217,34 +230,133 @@ func createProjectionForComplexAggregation(a *Aggregator, qp *QueryProjection) (
 	return p, nil
 }
 
-func createProjectionWithoutAggr(qp *QueryProjection, src ops.Operator) (*Projection, error) {
-	proj := &Projection{
-		Source: src,
-	}
-
-	for _, e := range qp.SelectExprs {
-		if _, isStar := e.Col.(*sqlparser.StarExpr); isStar {
-			return nil, errHorizonNotPlanned()
+func createProjectionWithoutAggr(ctx *plancontext.PlanningContext, qp *QueryProjection, src ops.Operator) (*Projection, error) {
+	// first we need to check if we have all columns or there are still unexpanded stars
+	aes, err := slice.MapWithError(qp.SelectExprs, func(from SelectExpr) (*sqlparser.AliasedExpr, error) {
+		ae, ok := from.Col.(*sqlparser.AliasedExpr)
+		if !ok {
+			return nil, fmt.Errorf("star found")
 		}
-		ae, err := e.GetAliasedExpr()
+		return ae, nil
+	})
 
+	if err != nil {
+		// if we have unexpanded expressions, we take this shortcut and hope we don't need any offsets from this plan
+		return newStarProjection(src, qp)
+	}
+
+	proj := newAliasedProjection(nil)
+	sqc := &SubQueryContainer{}
+	outerID := TableID(src)
+	for _, ae := range aes {
+		org := sqlparser.CloneRefOfAliasedExpr(ae)
+		expr := ae.Expr
+		newExpr, subqs, err := sqc.pullOutValueSubqueries(ctx, expr, outerID, false)
 		if err != nil {
 			return nil, err
 		}
-		expr := ae.Expr
-		if sqlparser.ContainsAggregation(expr) {
-			aggr, ok := expr.(sqlparser.AggrFunc)
-			if !ok {
-				// need to add logic to extract aggregations and pushed them to the top level
-				return nil, vterrors.VT12001(fmt.Sprintf("unsupported aggregation expression: %s", sqlparser.String(expr)))
+		if newExpr == nil {
+			// there was no subquery in this expression
+			_, err := proj.addUnexploredExpr(org, expr)
+			if err != nil {
+				return nil, err
 			}
-			expr = aggr.GetArg()
-			if expr == nil {
-				expr = sqlparser.NewIntLiteral("1")
+		} else {
+			err := proj.addSubqueryExpr(org, newExpr, subqs...)
+			if err != nil {
+				return nil, err
 			}
 		}
-
-		proj.addUnexploredExpr(ae, expr)
 	}
+	proj.Source = sqc.getRootOperator(src)
 	return proj, nil
 }
+
+func newStarProjection(src ops.Operator, qp *QueryProjection) (*Projection, error) {
+	cols := sqlparser.SelectExprs{}
+
+	for _, expr := range qp.SelectExprs {
+		err := sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+			_, isSubQ := node.(*sqlparser.Subquery)
+			if !isSubQ {
+				return true, nil
+			}
+			return false, vterrors.VT09015()
+		}, expr.Col)
+		if err != nil {
+			return nil, err
+		}
+		cols = append(cols, expr.Col)
+	}
+
+	return &Projection{
+		Source:  src,
+		Columns: StarProjections(cols),
+	}, nil
+}
+
+type subqueryExtraction struct {
```

Comment:
looking at this, I think we want to unify this with the builder parts of `SubqueryContainer`